### PR TITLE
plugin/kubernetes: Watch EndpointSlices

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. Why is this pull request needed and what does it do?

Watch EndpointSlice instead of Endpoints if the k8s server supports it.

I've done light manual testing on a kind cluster with a headless service that selects both ipv4 and ipv6 endpoints (in terms of service discovery, effectively a dual-stack headless service). I have not tested this on a "true" dual-stack cluster (but I don't think that matters in the context of just service discovery).

### 2. Which issues (if any) are related?

#4202

### 3. Which documentation changes (if any) need to be made?

probably none

### 4. Does this introduce a backward incompatible change or deprecation?

no

### CODE REVIEW

**Comment by chrisohaver** [comment]
I suspect the circleci tests might fail because it's running a version of k8s that does not support endpoint mirroring, but might support endpoint slices v1beta1.

**Comment by chrisohaver** [comment]
The last few commits add de-dup (where not already present) and tests for EndpointSlice skew address duplication.  With EndpointSlices, duplication of an Endpoint's Address is possible since we don't receive all EndpointSlices for any given Service at one time. https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#duplicate-endpoints

**Comment by chrisohaver** [comment]
> I have not tested this on a dual-stack cluster.

The [headless Service with custom Endpoints containing both IPv4 and IPv6 addresses](https://github.com/coredns/ci/blob/master/build/kubernetes/dns-test.yaml#L178-L204) in coredns/ci tests more or less covers this.  The [`SRV` test](https://github.com/coredns/ci/blob/master/test/kubernetes/srv_test.go#L73-L90) verifies that both IPv4 and IPv6 addresses of the headless Service are present in the response.

**Comment by chrisohaver** [comment]
EndpointSlices will be GA in K8s 1.20.  It would be cool if we are able get a version of CoreDNS with EndpointSlice support released in time to include with the K8s 1.20 release (code freeze Nov 12).

File: plugin/kubernetes/controller.go [hunk 1]
```diff
@@ -246,6 +272,62 @@ func namespaceListFunc(ctx context.Context, c kubernetes.Interface, s labels.Sel
...
+	return func(options meta.ListOptions) (watch.Interface, error) {
+		if s != nil {
+			options.LabelSelector = s.String()
+		}
+		w, err := c.CoreV1().Services(ns).Watch(ctx, options)
+		return w, err
+	}
+}
+
+func podWatchFunc(ctx context.Context, c kubernetes.Interface, ns string, s labels.Selector) func(options meta.ListOptions) (watch.Interface, error) {
```
Comment (bradbeam) [inline plugin/kubernetes/controller.go:286]: is this being used anywhere?

File: plugin/kubernetes/controller.go [hunk 2]
```diff
@@ -226,6 +242,16 @@ func podListFunc(ctx context.Context, c kubernetes.Interface, ns string, s label
 	}
 }
 
+func endpointSliceListFunc(ctx context.Context, c kubernetes.Interface, ns string, s labels.Selector) func(meta.ListOptions) (runtime.Object, error) {
+	return func(opts meta.ListOptions) (runtime.Object, error) {
+		if s != nil {
+			opts.LabelSelector = s.String()
+		}
+		listV1, err := c.DiscoveryV1beta1().EndpointSlices(ns).List(ctx, opts)
```
Comment (bradbeam) [inline plugin/kubernetes/controller.go]: nit, should we just `return c.DiscoveryV1beta1().EndpointSlices(ns).List(ctx, opts)`

File: plugin/kubernetes/controller.go (hunk 1 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go:286]: Yes.  It's used when we watch Pods in with the option `pods verified`.

File: plugin/kubernetes/controller.go (hunk 2 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go]: Did it in the other ListFunc and WatchFunc Functions too (for Services, Pods, and Endpoints).

**Review by rajansandeep — APPROVED** [review]
Thank you for adding this @chrisohaver 
This looks good to me!

File: plugin/kubernetes/controller.go [hunk 3]
```diff
@@ -130,16 +133,29 @@ func newdnsController(ctx context.Context, kubeClient kubernetes.Interface, opts
...
-			&cache.ListWatch{
-				ListFunc:  endpointsListFunc(ctx, dns.client, api.NamespaceAll, dns.selector),
-				WatchFunc: endpointsWatchFunc(ctx, dns.client, api.NamespaceAll, dns.selector),
-			},
-			&api.Endpoints{},
-			cache.ResourceEventHandlerFuncs{AddFunc: dns.Add, UpdateFunc: dns.Update, DeleteFunc: dns.Delete},
-			cache.Indexers{epNameNamespaceIndex: epNameNamespaceIndexFunc, epIPIndex: epIPIndexFunc},
-			object.DefaultProcessor(object.ToEndpoints(opts.skipAPIObjectsCleanup), dns.recordDNSProgrammingLatency),
-		)
+		if opts.useEndpointSlices {
```
Comment (miekg) [inline plugin/kubernetes/controller.go:142]: this needs to be simplified; seems `&api.EndpointSlice{}` is the only difference? That can be set latter in a small if-statement.

The endpoint slice also missed the dns latency for some reason?

Update: only the ListWatch seems difference, but the same argument holds

File: plugin/kubernetes/controller.go [hunk 4]
```diff
@@ -241,8 +263,56 @@ func namespaceListFunc(ctx context.Context, c kubernetes.Interface, s labels.Sel
 		if s != nil {
 			opts.LabelSelector = s.String()
 		}
-		listV1, err := c.CoreV1().Namespaces().List(ctx, opts)
-		return listV1, err
+		return c.CoreV1().Namespaces().List(ctx, opts)
+	}
+}
+
+func serviceWatchFunc(ctx context.Context, c kubernetes.Interface, ns string, s labels.Selector) func(options meta.ListOptions) (watch.Interface, error) {
```
Comment (miekg) [inline plugin/kubernetes/controller.go:277]: are these functions moved? Or why do we need these now?

File: plugin/kubernetes/handler_test.go [hunk 5]
```diff
@@ -643,8 +643,9 @@ var epsIndex = map[string][]*object.Endpoints{
 				},
 			},
 		},
-		Name:      "svc1",
+		Name:      "svc1-aaaaa",
```
Comment (miekg) [inline plugin/kubernetes/handler_test.go]: why is the -`aaaaa` added here?

File: plugin/kubernetes/kubernetes.go [hunk 6]
```diff
@@ -244,6 +245,17 @@ func (k *Kubernetes) InitKubeCache(ctx context.Context) (err error) {
 
 	k.opts.zones = k.Zones
 	k.opts.endpointNameMode = k.endpointNameMode
+	// Enable use of endpoint slices if the API supports the discovery v1 beta1 api
+	if _, err := kubeClient.Discovery().ServerResourcesForGroupVersion(discovery.SchemeGroupVersion.String()); err == nil {
+		k.opts.useEndpointSlices = true
```
Comment (miekg) [inline plugin/kubernetes/kubernetes.go:251]: log.Infof this?

File: plugin/kubernetes/kubernetes.go [hunk 7]
```diff
@@ -244,6 +245,17 @@ func (k *Kubernetes) InitKubeCache(ctx context.Context) (err error) {
...
+	}
+	// Disable use of endpoint slices for k8s versions 1.17 and 1.18. Endpoint slices were
+	// introduced in 1.17 but not fully integrated and enabled by default until k8s 1.19.
+	sv, _ := kubeClient.ServerVersion()
+	if strings.HasPrefix(sv.String(), "v1.18") || strings.HasPrefix(sv.String(), "v1.17") {
```
Comment (miekg) [inline plugin/kubernetes/kubernetes.go]: hasn't the version stuff of k8s a whole api to check this? Meaning sv.Major() and friends exist and can be checked here?

File: plugin/kubernetes/kubernetes.go [hunk 8]
```diff
@@ -455,6 +469,11 @@ func (k *Kubernetes) findServices(r recordRequest, zone string) (services []msg.
 							s := msg.Service{Host: addr.IP, Port: int(p.Port), TTL: k.ttl}
 							s.Key = strings.Join([]string{zonePath, Svc, svc.Namespace, svc.Name, endpointHostname(addr, k.endpointNameMode)}, "/")
 
+							if _, dup := uniq[s]; dup {
```
Comment (miekg) [inline plugin/kubernetes/kubernetes.go]: unclear why this is needed and I think we _alread_ remove duplicates from the answer

File: plugin/kubernetes/kubernetes_test.go [hunk 9]
```diff
@@ -151,22 +152,24 @@ func (APIConnServiceTest) EpIndex(string) []*object.Endpoints {
...
-						{IP: "172.0.0.3"},
+						{IP: "172.0.0.2"},
 					},
 					Ports: []object.EndpointPort{
 						{Port: 80, Protocol: "tcp", Name: "http"},
 					},
 				},
 			},
-			Name:      "hdls1",
+			Name:      "hdls1-bbbbb",
```
Comment (miekg) [inline plugin/kubernetes/kubernetes_test.go]: unclear what these suffixes mean and why they need to be added to existing test cases?

File: plugin/kubernetes/ns.go [hunk 10]
```diff
@@ -82,9 +88,10 @@ func (k *Kubernetes) nsAddrs(external bool, zone string) []dns.RR {
 		rr.Hdr.Name = svcNames[i]
 		rr.A = ip
 		rrs[i] = rr
+		i++
 	}
 
-	return rrs
+	return rrs[0:i]
```
Comment (miekg) [inline plugin/kubernetes/ns.go]: this can be shorted to `rrs[:i]` but I don't think we need this because uniqueness is done else where.

File: plugin/kubernetes/ns_test.go [hunk 11]
```diff
@@ -96,8 +98,23 @@ func (APIConnTest) EpIndexReverse(ip string) []*object.Endpoints {
 					},
 				},
 			},
-			Name:      "dns6-service",
+			Name:      "dns6-service-aaaaa",
```
Comment (miekg) [inline plugin/kubernetes/ns_test.go]: please make this meaningful or add a completely new test

File: plugin/kubernetes/object/endpoint.go [hunk 12]
```diff
@@ -56,6 +57,17 @@ func ToEndpoints(skipCleanup bool) ToFunc {
 	}
 }
 
+// SliceToEndpoints returns a function that converts an *discovery.EndpointSlice to a *Endpoints.
+func SliceToEndpoints(skipCleanup bool) ToFunc {
```
Comment (miekg) [inline plugin/kubernetes/object/endpoint.go]: inclined to say this should move to an `endpointslice.go` file - if these are actual objects in the k8s api, which I think they are?
function name should also be `EndpointSliceToEndpoints` to make super clear what being converrted.

File: plugin/kubernetes/object/endpoint.go [hunk 13]
```diff
@@ -108,6 +120,49 @@ func toEndpoints(skipCleanup bool, end *api.Endpoints) *Endpoints {
...
+			if end.TargetRef != nil {
+				ea.TargetRefName = end.TargetRef.Name
+			}
+			// EndpointSlice does not contain NodeName, leave blank
+			e.Subsets[0].Addresses = append(e.Subsets[0].Addresses, ea)
+			e.IndexIP = append(e.IndexIP, a)
+		}
+	}
+
+	if !skipCleanup {
```
Comment (miekg) [inline plugin/kubernetes/object/endpoint.go:159]: Only metrics tests seems to use this boolean?
~~~
 grep skipAPIObjectsCleanup -ri 
metrics_test.go:		skipAPIObjectsCleanup: true,
~~~
If so, please remove this from the API

File: plugin/kubernetes/xfr.go [hunk 14]
```diff
@@ -84,10 +84,6 @@ func (k *Kubernetes) Transfer(zone string, serial uint32) (<-chan []dns.RR, erro
 				endpointsList := k.APIConn.EpIndex(svc.Name + "." + svc.Namespace)
 
 				for _, ep := range endpointsList {
-					if ep.Name != svc.Name || ep.Namespace != svc.Namespace {
```
Comment (miekg) [inline plugin/kubernetes/xfr.go:87]: why this change?

File: plugin/kubernetes/object/endpoint.go (hunk 13 above)
Comment (miekg) [reply plugin/kubernetes/object/endpoint.go:159]: this can be done in a new PR

File: plugin/kubernetes/controller.go (hunk 3 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go:142]: Thanks for reviewing! Agreed, I can reduce some code there.

> The endpoint slice also missed the dns latency for some reason?

When I tested in a local cluster, I saw that EndpointSlices did not have the annotation (or equivalent other field) required to calculate latency (something that holds the date of the last update).  I'll double check in latest k8s release.

File: plugin/kubernetes/controller.go (hunk 4 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go:277]: They are not new, they are moved from `watch.go`, which was originally held an implementation of the since removed Watcher interface. At some point, these completely unrelated (except by name only) k8s api WatchFunc functions snuck in there.  When the CoreDNS watch feature was removed, the Watcher implementation was removed, and all that was left were the k8s api WatchFunc.

File: plugin/kubernetes/handler_test.go (hunk 5 above)
Comment (chrisohaver) [reply plugin/kubernetes/handler_test.go]: This is what endpointSlice names look like... by convention: the service name, dash, 5 random alpha-nums.
Having this reflected in the tests that we do not have assumptions in the code that and endpoint name = service name, which is no longer the case for EndpointSlice.

File: plugin/kubernetes/kubernetes.go (hunk 8 above)
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: IIRC, the backend does duplicate removal for some query types, but not all (e.g. not for A/AAAA).

I added de-dup (where not already present) and tests for EndpointSlice skew address duplication. With EndpointSlices, duplication of an Endpoint's Address is possible since we don't receive all EndpointSlices for any given Service at one time. https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#duplicate-endpoints

File: plugin/kubernetes/kubernetes_test.go (hunk 9 above)
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes_test.go]: This is what endpointSlice names look like... by convention: the service name, dash, 5 random alpha-nums.
Having this reflected in the tests that we do not have assumptions in the code that and endpoint name = service name, which is no longer the case for EndpointSlice.

File: plugin/kubernetes/xfr.go (hunk 14 above)
Comment (aojea) [reply plugin/kubernetes/xfr.go:87]: I assume that this is because previously there was a 1:1 relation service - endpoint, referenced by name.
with endpoint slices service to slices is 1:N and the slices reference the service by label
xref: https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#ownership

Comment (chrisohaver) [reply plugin/kubernetes/xfr.go:87]: It was a pointless check. Name and namespace would always match, since we are getting the list from an index lookup on name and namespace. It was probably copy-pasted from `findServices()` which does need to check due to wildcard queries. 
Anyways, now comparing object.Endpoint.Name to object.Service.Name is no longer valid, since the object.Endpoint could have been derived from an api.EndpointSlice instead of a api.Endpoint.

File: plugin/kubernetes/object/endpoint.go (hunk 12 above)
Comment (chrisohaver) [reply plugin/kubernetes/object/endpoint.go]: EndpointSlices _are_ separate objects, but to simplify supporting both Endpoint and EndpointSlice in the code at the same time, I am distilling both Endpoints and EndpointSlice into object.Endpoint, instead of adding a new struct called object.EndpointSlice.  Since the new conversion functions work with the existing object.Endpoints struct, I put them in `object/endpoints.go`, but I can move them to a separate file if that makes more sense.

> function name should also be EndpointSliceToEndpoints to make super clear what being converrted.

Yeah - I can change that.

**Comment by aojea** [comment]
@chrisohaver unrelated to this PR, but taking the opportunity to comment because it will affect CoreDNS.
Something new in tKuberntes dual-stack is that Services will pluralize the ClusterIP, so dual-stack services may also have ClusterIPs as an array with one IPv4 and one IPv6 address

**Comment by chrisohaver** [comment]
> @chrisohaver unrelated to this PR, but taking the opportunity to comment because it will affect CoreDNS.
> Something new in tKuberntes dual-stack is that Services will pluralize the ClusterIP, so dual-stack services may also have ClusterIPs as an array with one IPv4 and one IPv6 address

@aojea, OK thanks (IIRC, I recently reviewed the PR that added this). Has that landed in a released version of the API yet?

**Comment by aojea** [comment]
> @aojea, OK thanks (IIRC, I recently reviewed the PR that added this). Has that landed in a released version of the API yet?

it landed yesterday https://github.com/kubernetes/kubernetes/pull/91824, 
but will be under a feature gate in alpha and not enabled by default in 1.20

**Comment by chrisohaver** [comment]
> > @aojea, OK thanks (IIRC, I recently reviewed the PR that added this). Has that landed in a released version of the API yet?
> 
> it landed yesterday [kubernetes/kubernetes#91824](https://github.com/kubernetes/kubernetes/pull/91824),
> but will be under a feature gate in alpha and not enabled by default in 1.20

OK.  IIUC, we shouldn't add support for _that_ to a CoreDNS release until a version of the K8s API is released with that in it. Otherwise I think we'd need to release a CoreDNS which was compiled against master k8s-api or a pre-release tag.

**Comment by chrisohaver** [comment]
A general note: Supporting both Endpoints and EndpointSlices in the codebase can be relatively short lived.  When K8s 1.21 is released, the 3 supported k8s would be 1.21, 1.20, 1.19, all of which fully support EndpointSlices. So IMO, we can drop the Endpoints watch code with the version of CoreDNS that is released with K8s 1.21.

**Comment by johnbelamaric** [comment]
Support is now 12 months not 9, so typically that means 4 releases. FYI.

File: plugin/kubernetes/ns_test.go (hunk 11 above)
Comment (chrisohaver) [reply plugin/kubernetes/ns_test.go]: Thanks. I've refactored, added comments and made the test data more compact and (i think) easier to read.  The test I had modified wasn't testing the right thing, and fixing it exposed a bug in nsAddrs, which is ~not~ _now_ fixed.

File: plugin/kubernetes/controller.go (hunk 3 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go:142]: ~Ah, I found an updated timestamp in another place in the EndpointSlice object, I'll adjust the code to use it.~

No, OK , I see now that the same annotation _is_ added when EndpointSlices are created by the EndpointSliceController, but not when created by the EndpointSliceMirrorController.  So the existing code should work, with some generification to handle both Endpoints and EndpointSlice.

Comment (miekg) [reply plugin/kubernetes/controller.go:142]: listWatch can be moved out of the if-then, and just assigned, making this more readeble

File: plugin/kubernetes/controller.go [hunk 15]
```diff
@@ -130,15 +133,35 @@ func newdnsController(ctx context.Context, kubeClient kubernetes.Interface, opts
...
+			toFuncFunc        func(bool) object.ToFunc
+			recordLatencyFunc object.RecordLatencyFunc
+		)
+		if opts.useEndpointSlices {
+			apiObj = &discovery.EndpointSlice{}
+			listWatch = &cache.ListWatch{
+				ListFunc:  endpointSliceListFunc(ctx, dns.client, api.NamespaceAll, dns.selector),
+				WatchFunc: endpointSliceWatchFunc(ctx, dns.client, api.NamespaceAll, dns.selector),
+			}
+			toFuncFunc = object.EndpointSliceToEndpoints
```
Comment (miekg) [inline plugin/kubernetes/controller.go]: why FuncFunc?

File: plugin/kubernetes/handler_test.go (hunk 5 above)
Comment (miekg) [reply plugin/kubernetes/handler_test.go]: really, wow? Please add a comment with ref to some docs, this looks like someone was just banging the keyboard, not a conscious design decision.

File: plugin/kubernetes/kubernetes.go (hunk 8 above)
Comment (miekg) [reply plugin/kubernetes/kubernetes.go]: the dedup exists in backend.go (other discussion if that is still useful), but we should add it there, not here.
Thanks the ref to those docs

File: plugin/kubernetes/object/endpoint.go (hunk 13 above)
Comment (miekg) [reply plugin/kubernetes/object/endpoint.go:159]: https://github.com/coredns/coredns/pull/4244

File: plugin/kubernetes/xfr.go (hunk 14 above)
Comment (miekg) [reply plugin/kubernetes/xfr.go:87]: ack ,thanks

File: plugin/kubernetes/handler_test.go (hunk 5 above)
Comment (chrisohaver) [reply plugin/kubernetes/handler_test.go]: I'll make them just incremental numbers.  The random string is the name scheme that K8s controllers use when creating endpoints (similar to how Pods are named when deployed per Deployments). I was just imitating that to make the data more "real", but it's not necessary at all.  All that matters is that they don't match the service name.

File: plugin/kubernetes/controller.go (hunk 15 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go]: It's a function that returns function that returns a function.  IOW, a function that returns an object.ToFunc, not an object.ToFunc itself. Made sense to me, but I'll rename it to just `to`.

File: plugin/kubernetes/kubernetes.go (hunk 8 above)
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: Yes, the backend already does de-dup for all types that I can see. I was staring right at it and didn't see it.  I've removed all de-dup from this PR.

File: plugin/kubernetes/kubernetes.go (hunk 6 above)
Comment (miekg) [reply plugin/kubernetes/kubernetes.go:251]: is this done? I want to resolve my remaining comments, so I can lgtm this

Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go:251]: Yes. A message is logged when disabling endpointslice watch for k8s before 1.19.

**Review by miekg — APPROVED** [review]
looks good enough

**Comment by pires** [comment]
It seems this PR actually introduces a breaking change as CoreDNS will not report `kubernetes` plug-in is ready if it's not given permissions to watch `EndpointSlice`.

I just reproduced this w/ a new cluster provisioned w/ kubeadm 1.19.10.

**Comment by aojea** [comment]
I have a PR for fixing it, kubeadm should provide the permissions for slices

**Comment by chrisohaver** [comment]
> It seems this PR actually introduces a breaking change

~@pires, Correct.  This is fixed in #4580.~

Correction: This PR did introduce a breaking change, but not the one you refer to.

> CoreDNS will not report kubernetes plug-in is ready if it's not given permissions to watch EndpointSlice

This is intended behavior.  CoreDNS needs to have access to endpointslice to work in k8s >= 1.19.

**Comment by pires** [comment]
>CoreDNS needs to have access to endpointslice to work in k8s >= 1.19.

I can't find any reference to this in the release docs for any of the 1.8.x patches. So let's agree to disagree.

Thank you, @aojea for addressing this on the kubeadm side.

**Comment by chrisohaver** [comment]
@pires, https://github.com/coredns/deployment/pull/246

**Comment by chrisohaver** [comment]
In retrospect, we should have bumped the minor version and documented this as a backward incompatible deprecation. 

... I've added it to the 1.8.1 notes retroactively, FWIW.

**Comment by pires** [comment]
Thank you!
